File: sort/QuickSort.java
import java.util.Arrays;

class QuickSort {


    public static void main(String[] args) {
        int[] arr = new int[]{5,2,18,3,11,90,45,22};
        quickSort(arr, 0, arr.length - 1);
        System.out.println(Arrays.toString(arr));
    }


    public static void quickSort(int[] arr, int start, int end) {
        if (arr == null || arr.length < 2 || start >= end) {
            return;
        }

        int i = start, j = end;
        int pivot = arr[start + (end - start) / 2];

        while (i <= j) {
            while (arr[i] < pivot) {
                i++;
            }
            while (arr[j] > pivot) {  // 使用>=时，会产生java.lang.StackOverflowError
                j--;
            }
            if (i < j) {
                int tmp = arr[j];
                arr[j] = arr[i];
                arr[i] = tmp;
                i++;
                j--;
            } else if (i == j) {
                i++;
            }
        }
        quickSort(arr, start, j);
        quickSort(arr, i, end);
    }


}

File: sort/quick_sort.cc
/**
 * quick sort
 * 快速排序
 */

#include <iostream>
#include <vector>
#include <time.h>
#include <cstdlib>

void quickSort(std::vector<int> &arr, int start, int end) {
    if (arr.size() <= 1 || start >= end) {
        return;
    }

    int pivot = arr[start + (end - start) / 2];     // 这里要注意应该是数组的值
    int i = start, j = end;

    while (i <= j) {
        while (arr[i] < pivot) {
            i++;
        }
        while (arr[j] > pivot) {
            j--;
        }
        if (i < j) {
            int tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
            i++;
            j--;
        } else if (i == j) {      // 这里要注意判断i==j
            i++;
        }
    }
    quickSort(arr, start, j);
    quickSort(arr, i, end);
}

void print(std::vector<int> &arr){
    for (int i = 0; i < arr.size(); i++) {
        std::cout << arr[i] << " ";
    }
    std::cout << std::endl;
}

int main() {
    std::vector<int> arr;
    srand((unsigned) time(NULL));
    for (int i = 0; i < 10; i++) {
        arr.push_back(rand() % 100);
    }
    print(arr);
    quickSort(arr, 0, arr.size() - 1);
    print(arr);

    return 0;
}

